pkg/alerter/alerter.go

```diff
@@ -0,0 +1,245 @@
+
+	dashboards := make(chan *configpb.Dashboard)
+	var wg sync.WaitGroup
+
+	errCh := make(chan error)
+
+	// Launch monitoring goroutines
+	for i := 0; i < concurrency; i++ {
+		wg.Add(1)
+		go func() {
+			for dash := range dashboards {
+				// Read the new summary file from the expected location (summary)
+				// If it's not there, alerts can't be generated for this dashboard
+				newPath, err := path.ResolveReference(&url.URL{Path: summaryPath(dash.Name)})
+				if err != nil {
+					logrus.Errorf("Could not resolve path to current summaries of %s", dash.Name)
+					errCh <- err
+					continue
+				}
+
+				newSummary, err := ReadSummary(ctx, client, *newPath)
+				if err != nil {
+					logrus.Errorf("Could not read summary from %v", newPath)
+					errCh <- err
+					continue
+				}
+
+				// Read the old summary file from the expected location (alerter)
+				// It may not be there the first time this is run
+				oldPath, err := path.ResolveReference(&url.URL{Path: alerterPath(dash.Name)})
+				if err != nil {
+					logrus.Errorf("Could not resolve path to previous summaries of %s", dash.Name)
+					errCh <- err
+					continue
+				}
+
+				oldSummary, err := ReadSummary(ctx, client, *oldPath)
+				if err != nil {
+					logrus.Warnf("Could not read summary from %v; trying write", oldPath)
+				}
+
+				if !confirm {
+					logrus.Infof("Summarizers read for dashboard %s (dry-run)", dash.Name)
+					continue
+				}
+
+				// Detect changes and call the Klaxon function(s) if those changes are pertinent
+				if oldSummary != nil {
+					callOutDifferences(ctx, newSummary, oldSummary, dash, subscribers)
+				}
+
+				// Write the new summary to where the old summary was
+				err = summarizer.WriteSummary(ctx, client, *oldPath, newSummary)
+				if err != nil {
+					logrus.Errorf("Could not write summary to %v", oldPath)
+					errCh <- err
+					continue
+				}
+				errCh <- nil
+			}
+			wg.Done()
+		}()
+	}
+
+	// Subroutine to concatenate errors together
+	resultCh := make(chan error)
+	go func() {
+		var errs []string
+		for err := range errCh {
+			if err == nil {
+				continue
+			}
+			errs = append(errs, err.Error())
+		}
+		if n := len(errs); n > 0 {
+			resultCh <- fmt.Errorf("failed to update %d dashboards: %v", n, strings.Join(errs, ", "))
+		}
+		resultCh <- nil
+		close(resultCh)
+	}()
+
+	// Queue up all the dashboards
+	for _, d := range cfg.Dashboards {
+		if dashboard != "" && dashboard != d.Name {
+			logrus.WithField("dashboard", d.Name).Info("Skipping")
+			continue
+		}
+		dashboards <- d
+	}
+	close(dashboards)
+	wg.Wait() //wait for them to stop reporting
+	close(errCh)
+	return <-resultCh
+}
+
+// Paths to relevant GCS files
+var (
+	normalizer = regexp.MustCompile(`[^a-z0-9]+`)
+)
+
+// A normalized path for a summary file is "prefix-dashboardnameinlowercase"
+func summaryPath(name string) string {
+	return "summary-" + normalizer.ReplaceAllString(strings.ToLower(name), "")
```

Comment:
Seems like this path ought to come from the summarizer package, as it is a detail of the summarizer